// modelfusion/model-function/transcribe-speech/transcribe.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transcribe = void 0;
const executeCall_js_1 = require("../executeCall.cjs");
const ModelFunctionPromise_js_1 = require("../ModelFunctionPromise.cjs");
/**
 * Transcribe audio data into text.
 *
 * @example
 * const data = await fs.promises.readFile("data/test.mp3");
 *
 * const transcription = await transcribe(
 *   new OpenAITranscriptionModel({ model: "whisper-1" }),
 *   {
 *     type: "mp3",
 *     data,
 *   }
 * );
 */
function transcribe(model, data, options) {
    return new ModelFunctionPromise_js_1.ModelFunctionPromise((0, executeCall_js_1.executeCall)({
        functionType: "transcription",
        input: data,
        model,
        options,
        generateResponse: async (options) => {
            const result = await model.doTranscribe(data, options);
            return {
                response: result.response,
                extractedValue: result.transcription,
            };
        },
    }));
}
exports.transcribe = transcribe;
