// modelfusion/retriever/retrieve.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.retrieve = void 0;
const nanoid_1 = require("nanoid");
const FunctionEventSource_js_1 = require("../core/FunctionEventSource.cjs");
const GlobalFunctionLogging_js_1 = require("../core/GlobalFunctionLogging.cjs");
const GlobalFunctionObservers_js_1 = require("../core/GlobalFunctionObservers.cjs");
const AbortError_js_1 = require("../core/api/AbortError.cjs");
const getFunctionCallLogger_js_1 = require("../core/getFunctionCallLogger.cjs");
const getRun_js_1 = require("../core/getRun.cjs");
const DurationMeasurement_js_1 = require("../util/DurationMeasurement.cjs");
const runSafe_js_1 = require("../util/runSafe.cjs");
async function retrieve(retriever, query, options) {
    const run = await (0, getRun_js_1.getRun)(options?.run);
    const eventSource = new FunctionEventSource_js_1.FunctionEventSource({
        observers: [
            ...(0, getFunctionCallLogger_js_1.getFunctionCallLogger)(options?.logging ?? (0, GlobalFunctionLogging_js_1.getGlobalFunctionLogging)()),
            ...(0, GlobalFunctionObservers_js_1.getGlobalFunctionObservers)(),
            ...(run?.functionObserver != null ? [run.functionObserver] : []),
            ...(options?.observers ?? []),
        ],
        errorHandler: run?.errorHandler,
    });
    const durationMeasurement = (0, DurationMeasurement_js_1.startDurationMeasurement)();
    const startMetadata = {
        functionType: "retrieve",
        callId: `call-${(0, nanoid_1.nanoid)()}`,
        runId: run?.runId,
        sessionId: run?.sessionId,
        userId: run?.userId,
        functionId: options?.functionId,
        query,
        timestamp: durationMeasurement.startDate,
        startTimestamp: durationMeasurement.startDate,
    };
    eventSource.notify({
        eventType: "started",
        ...startMetadata,
    });
    const result = await (0, runSafe_js_1.runSafe)(() => retriever.retrieve(query, {
        functionId: options?.functionId,
        logging: options?.logging,
        observers: options?.observers,
        run,
    }));
    const finishMetadata = {
        eventType: "finished",
        ...startMetadata,
        finishTimestamp: new Date(),
        durationInMs: durationMeasurement.durationInMs,
    };
    if (!result.ok) {
        if (result.isAborted) {
            eventSource.notify({
                ...finishMetadata,
                eventType: "finished",
                result: {
                    status: "abort",
                },
            });
            throw new AbortError_js_1.AbortError();
        }
        eventSource.notify({
            ...finishMetadata,
            eventType: "finished",
            result: {
                status: "error",
                error: result.error,
            },
        });
        throw result.error;
    }
    eventSource.notify({
        ...finishMetadata,
        eventType: "finished",
        result: {
            status: "success",
            value: result.value,
        },
    });
    return result.value;
}
exports.retrieve = retrieve;
