// modelfusion/model-provider/stability/StabilityError.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.failedStabilityCallResponseHandler = exports.StabilityError = exports.stabilityErrorDataSchema = void 0;
const zod_1 = require("zod");
const ApiCallError_js_1 = require("../../core/api/ApiCallError.cjs");
const parseJSON_js_1 = require("../../util/parseJSON.cjs");
exports.stabilityErrorDataSchema = zod_1.z.object({
    message: zod_1.z.string(),
});
class StabilityError extends ApiCallError_js_1.ApiCallError {
    constructor({ data, statusCode, url, requestBodyValues, message = data.message, }) {
        super({ message, statusCode, requestBodyValues, url });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.data = data;
    }
}
exports.StabilityError = StabilityError;
const failedStabilityCallResponseHandler = async ({ response, url, requestBodyValues }) => {
    const responseBody = await response.text();
    const parsedError = (0, parseJSON_js_1.parseJsonWithZod)(responseBody, exports.stabilityErrorDataSchema);
    return new StabilityError({
        url,
        requestBodyValues,
        statusCode: response.status,
        data: parsedError,
    });
};
exports.failedStabilityCallResponseHandler = failedStabilityCallResponseHandler;
