// modelfusion/model-provider/stability/StabilityImageGenerationModel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StabilityImageGenerationModel = void 0;
const zod_1 = require("zod");
const callWithRetryAndThrottle_js_1 = require("../../core/api/callWithRetryAndThrottle.cjs");
const postToApi_js_1 = require("../../core/api/postToApi.cjs");
const AbstractModel_js_1 = require("../../model-function/AbstractModel.cjs");
const PromptFormatImageGenerationModel_js_1 = require("../../model-function/generate-image/PromptFormatImageGenerationModel.cjs");
const StabilityApiConfiguration_js_1 = require("./StabilityApiConfiguration.cjs");
const StabilityError_js_1 = require("./StabilityError.cjs");
const StabilityImageGenerationPrompt_js_1 = require("./StabilityImageGenerationPrompt.cjs");
/**
 * Create an image generation model that calls the Stability AI image generation API.
 *
 * @see https://api.stability.ai/docs#tag/v1generation/operation/textToImage
 *
 * @example
 * const image = await generateImage(
 *   new StabilityImageGenerationModel({
 *     model: "stable-diffusion-512-v2-1",
 *     cfgScale: 7,
 *     clipGuidancePreset: "FAST_BLUE",
 *     height: 512,
 *     width: 512,
 *     samples: 1,
 *     steps: 30,
 *   })
 *   [
 *     { text: "the wicked witch of the west" },
 *     { text: "style of early 19th century painting", weight: 0.5 },
 *   ]
 * );
 */
class StabilityImageGenerationModel extends AbstractModel_js_1.AbstractModel {
    constructor(settings) {
        super({ settings });
        Object.defineProperty(this, "provider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "stability"
        });
    }
    get modelName() {
        return this.settings.model;
    }
    async callAPI(input, options) {
        return (0, callWithRetryAndThrottle_js_1.callWithRetryAndThrottle)({
            retry: this.settings.api?.retry,
            throttle: this.settings.api?.throttle,
            call: async () => callStabilityImageGenerationAPI({
                ...this.settings,
                abortSignal: options?.run?.abortSignal,
                engineId: this.settings.model,
                textPrompts: input,
            }),
        });
    }
    get settingsForEvent() {
        const eventSettingProperties = [
            "baseUrl",
            "height",
            "width",
            "cfgScale",
            "clipGuidancePreset",
            "sampler",
            "samples",
            "seed",
            "steps",
            "stylePreset",
        ];
        return Object.fromEntries(Object.entries(this.settings).filter(([key]) => eventSettingProperties.includes(key)));
    }
    async doGenerateImage(prompt, options) {
        const response = await this.callAPI(prompt, options);
        return {
            response,
            base64Image: response.artifacts[0].base64,
        };
    }
    withBasicPrompt() {
        return this.withPromptFormat((0, StabilityImageGenerationPrompt_js_1.mapBasicPromptToStabilityFormat)());
    }
    withPromptFormat(promptFormat) {
        return new PromptFormatImageGenerationModel_js_1.PromptFormatImageGenerationModel({
            model: this,
            promptFormat,
        });
    }
    withSettings(additionalSettings) {
        return new StabilityImageGenerationModel(Object.assign({}, this.settings, additionalSettings));
    }
}
exports.StabilityImageGenerationModel = StabilityImageGenerationModel;
const stabilityImageGenerationModels = [
    "stable-diffusion-v1-5",
    "stable-diffusion-512-v2-1",
    "stable-diffusion-xl-1024-v0-9",
    "stable-diffusion-xl-1024-v1-0",
];
const stabilityImageGenerationResponseSchema = zod_1.z.object({
    artifacts: zod_1.z.array(zod_1.z.object({
        base64: zod_1.z.string(),
        seed: zod_1.z.number(),
        finishReason: zod_1.z.enum(["SUCCESS", "ERROR", "CONTENT_FILTERED"]),
    })),
});
async function callStabilityImageGenerationAPI({ api = new StabilityApiConfiguration_js_1.StabilityApiConfiguration(), abortSignal, engineId, height, width, textPrompts, cfgScale, clipGuidancePreset, sampler, samples, seed, steps, stylePreset, }) {
    return (0, postToApi_js_1.postJsonToApi)({
        url: api.assembleUrl(`/generation/${engineId}/text-to-image`),
        headers: api.headers,
        body: {
            height,
            width,
            text_prompts: textPrompts,
            cfg_scale: cfgScale,
            clip_guidance_preset: clipGuidancePreset,
            sampler,
            samples,
            seed,
            steps,
            style_preset: stylePreset,
        },
        failedResponseHandler: StabilityError_js_1.failedStabilityCallResponseHandler,
        successfulResponseHandler: (0, postToApi_js_1.createJsonResponseHandler)(stabilityImageGenerationResponseSchema),
        abortSignal,
    });
}
