// modelfusion/model-provider/stability/StabilityError.js
import { z } from "zod";
import { ApiCallError } from "../../core/api/ApiCallError.js";
import { parseJsonWithZod } from "../../util/parseJSON.js";
export const stabilityErrorDataSchema = z.object({
    message: z.string(),
});
export class StabilityError extends ApiCallError {
    constructor({ data, statusCode, url, requestBodyValues, message = data.message, }) {
        super({ message, statusCode, requestBodyValues, url });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.data = data;
    }
}
export const failedStabilityCallResponseHandler = async ({ response, url, requestBodyValues }) => {
    const responseBody = await response.text();
    const parsedError = parseJsonWithZod(responseBody, stabilityErrorDataSchema);
    return new StabilityError({
        url,
        requestBodyValues,
        statusCode: response.status,
        data: parsedError,
    });
};
