// modelfusion/core/DefaultRun.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultRun = void 0;
const nanoid_1 = require("nanoid");
const SuccessfulModelCall_js_1 = require("../model-function/SuccessfulModelCall.cjs");
const FunctionEventSource_js_1 = require("./FunctionEventSource.cjs");
class DefaultRun {
    constructor({ runId = `run-${(0, nanoid_1.nanoid)()}`, sessionId, userId, abortSignal, observers, errorHandler, } = {}) {
        Object.defineProperty(this, "runId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "sessionId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "userId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "abortSignal", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "errorHandler", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "events", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "functionEventSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "functionObserver", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {
                onFunctionEvent: (event) => {
                    this.events.push(event);
                    this.functionEventSource.notify(event);
                },
            }
        });
        this.runId = runId;
        this.sessionId = sessionId;
        this.userId = userId;
        this.abortSignal = abortSignal;
        this.errorHandler = errorHandler ?? ((error) => console.error(error));
        this.functionEventSource = new FunctionEventSource_js_1.FunctionEventSource({
            observers: observers ?? [],
            errorHandler: this.errorHandler.bind(this),
        });
    }
    get successfulModelCalls() {
        return (0, SuccessfulModelCall_js_1.extractSuccessfulModelCalls)(this.events);
    }
}
exports.DefaultRun = DefaultRun;
