// modelfusion/core/api/ApiCallError.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiCallError = void 0;
class ApiCallError extends Error {
    constructor({ message, url, requestBodyValues, statusCode, cause, isRetryable = statusCode === 429 || statusCode >= 500, }) {
        super(message);
        Object.defineProperty(this, "url", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "requestBodyValues", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "statusCode", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "cause", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "isRetryable", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = "ApiCallError";
        this.url = url;
        this.requestBodyValues = requestBodyValues;
        this.statusCode = statusCode;
        this.cause = cause;
        this.isRetryable = isRetryable;
    }
}
exports.ApiCallError = ApiCallError;
