// modelfusion/core/api/postToApi.js
import { ApiCallError } from "./ApiCallError.js";
export const createJsonResponseHandler = (responseSchema) => async ({ response, url, requestBodyValues }) => {
    const parsedResult = responseSchema.safeParse(await response.json());
    if (!parsedResult.success) {
        throw new ApiCallError({
            message: "Invalid JSON response",
            cause: parsedResult.error,
            statusCode: response.status,
            url,
            requestBodyValues,
        });
    }
    return parsedResult.data;
};
export const createTextResponseHandler = () => async ({ response }) => response.text();
export const createAudioMpegResponseHandler = () => async ({ response, url, requestBodyValues }) => {
    if (response.headers.get("Content-Type") !== "audio/mpeg") {
        throw new ApiCallError({
            message: "Invalid Content-Type (must be audio/mpeg)",
            statusCode: response.status,
            url,
            requestBodyValues,
        });
    }
    const arrayBuffer = await response.arrayBuffer();
    return Buffer.from(arrayBuffer);
};
export const postJsonToApi = async ({ url, headers, body, failedResponseHandler, successfulResponseHandler, abortSignal, }) => postToApi({
    url,
    headers: {
        ...headers,
        "Content-Type": "application/json",
    },
    body: {
        content: JSON.stringify(body),
        values: body,
    },
    failedResponseHandler,
    successfulResponseHandler,
    abortSignal,
});
export const postToApi = async ({ url, headers = {}, body, successfulResponseHandler, failedResponseHandler, abortSignal, }) => {
    try {
        const response = await fetch(url, {
            method: "POST",
            headers,
            body: body.content,
            signal: abortSignal,
        });
        if (!response.ok) {
            try {
                throw await failedResponseHandler({
                    response,
                    url,
                    requestBodyValues: body.values,
                });
            }
            catch (error) {
                if (error instanceof Error) {
                    if (error.name === "AbortError" || error instanceof ApiCallError) {
                        throw error;
                    }
                }
                throw new ApiCallError({
                    message: "Failed to process error response",
                    cause: error,
                    statusCode: response.status,
                    url,
                    requestBodyValues: body.values,
                });
            }
        }
        try {
            return await successfulResponseHandler({
                response,
                url,
                requestBodyValues: body.values,
            });
        }
        catch (error) {
            if (error instanceof Error) {
                if (error.name === "AbortError" || error instanceof ApiCallError) {
                    throw error;
                }
            }
            throw new ApiCallError({
                message: "Failed to process successful response",
                cause: error,
                statusCode: response.status,
                url,
                requestBodyValues: body.values,
            });
        }
    }
    catch (error) {
        if (error instanceof Error) {
            if (error.name === "AbortError") {
                throw error;
            }
        }
        // unwrap original error when fetch failed (for easier debugging):
        if (error instanceof TypeError && error.message === "fetch failed") {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            if (error.cause != null) {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                throw error.cause;
            }
        }
        throw error;
    }
};
