// modelfusion/core/api/postToApi.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.postToApi = exports.postJsonToApi = exports.createAudioMpegResponseHandler = exports.createTextResponseHandler = exports.createJsonResponseHandler = void 0;
const ApiCallError_js_1 = require("./ApiCallError.cjs");
const createJsonResponseHandler = (responseSchema) => async ({ response, url, requestBodyValues }) => {
    const parsedResult = responseSchema.safeParse(await response.json());
    if (!parsedResult.success) {
        throw new ApiCallError_js_1.ApiCallError({
            message: "Invalid JSON response",
            cause: parsedResult.error,
            statusCode: response.status,
            url,
            requestBodyValues,
        });
    }
    return parsedResult.data;
};
exports.createJsonResponseHandler = createJsonResponseHandler;
const createTextResponseHandler = () => async ({ response }) => response.text();
exports.createTextResponseHandler = createTextResponseHandler;
const createAudioMpegResponseHandler = () => async ({ response, url, requestBodyValues }) => {
    if (response.headers.get("Content-Type") !== "audio/mpeg") {
        throw new ApiCallError_js_1.ApiCallError({
            message: "Invalid Content-Type (must be audio/mpeg)",
            statusCode: response.status,
            url,
            requestBodyValues,
        });
    }
    const arrayBuffer = await response.arrayBuffer();
    return Buffer.from(arrayBuffer);
};
exports.createAudioMpegResponseHandler = createAudioMpegResponseHandler;
const postJsonToApi = async ({ url, headers, body, failedResponseHandler, successfulResponseHandler, abortSignal, }) => (0, exports.postToApi)({
    url,
    headers: {
        ...headers,
        "Content-Type": "application/json",
    },
    body: {
        content: JSON.stringify(body),
        values: body,
    },
    failedResponseHandler,
    successfulResponseHandler,
    abortSignal,
});
exports.postJsonToApi = postJsonToApi;
const postToApi = async ({ url, headers = {}, body, successfulResponseHandler, failedResponseHandler, abortSignal, }) => {
    try {
        const response = await fetch(url, {
            method: "POST",
            headers,
            body: body.content,
            signal: abortSignal,
        });
        if (!response.ok) {
            try {
                throw await failedResponseHandler({
                    response,
                    url,
                    requestBodyValues: body.values,
                });
            }
            catch (error) {
                if (error instanceof Error) {
                    if (error.name === "AbortError" || error instanceof ApiCallError_js_1.ApiCallError) {
                        throw error;
                    }
                }
                throw new ApiCallError_js_1.ApiCallError({
                    message: "Failed to process error response",
                    cause: error,
                    statusCode: response.status,
                    url,
                    requestBodyValues: body.values,
                });
            }
        }
        try {
            return await successfulResponseHandler({
                response,
                url,
                requestBodyValues: body.values,
            });
        }
        catch (error) {
            if (error instanceof Error) {
                if (error.name === "AbortError" || error instanceof ApiCallError_js_1.ApiCallError) {
                    throw error;
                }
            }
            throw new ApiCallError_js_1.ApiCallError({
                message: "Failed to process successful response",
                cause: error,
                statusCode: response.status,
                url,
                requestBodyValues: body.values,
            });
        }
    }
    catch (error) {
        if (error instanceof Error) {
            if (error.name === "AbortError") {
                throw error;
            }
        }
        // unwrap original error when fetch failed (for easier debugging):
        if (error instanceof TypeError && error.message === "fetch failed") {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            if (error.cause != null) {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                throw error.cause;
            }
        }
        throw error;
    }
};
exports.postToApi = postToApi;
