// modelfusion/model-provider/elevenlabs/ElevenLabsSpeechSynthesisModel.js
import { callWithRetryAndThrottle } from "../../core/api/callWithRetryAndThrottle.js";
import { createAudioMpegResponseHandler, postJsonToApi, } from "../../core/api/postToApi.js";
import { AbstractModel } from "../../model-function/AbstractModel.js";
import { ElevenLabsApiConfiguration } from "./ElevenLabsApiConfiguration.js";
import { failedElevenLabsCallResponseHandler } from "./ElevenLabsError.js";
/**
 * Synthesize speech using the ElevenLabs Text to Speech API.
 *
 * @see https://api.elevenlabs.io/docs#/text-to-speech/Text_to_speech_v1_text_to_speech__voice_id__post
 */
export class ElevenLabsSpeechSynthesisModel extends AbstractModel {
    constructor(settings) {
        super({ settings });
        Object.defineProperty(this, "provider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "elevenlabs"
        });
    }
    get modelName() {
        return this.settings.voice;
    }
    async callAPI(text, options) {
        return callWithRetryAndThrottle({
            retry: this.settings.api?.retry,
            throttle: this.settings.api?.throttle,
            call: async () => callElevenLabsTextToSpeechAPI({
                api: this.settings.api,
                abortSignal: options?.run?.abortSignal,
                text,
                voiceId: this.settings.voice,
                modelId: this.settings.model,
                voiceSettings: this.settings.voiceSettings,
            }),
        });
    }
    get settingsForEvent() {
        return {
            model: this.settings.model,
            voice: this.settings.voice,
            voiceSettings: this.settings.voiceSettings,
        };
    }
    generateSpeechResponse(text, options) {
        return this.callAPI(text, options);
    }
    withSettings(additionalSettings) {
        return new ElevenLabsSpeechSynthesisModel({
            ...this.settings,
            ...additionalSettings,
        });
    }
}
async function callElevenLabsTextToSpeechAPI({ api = new ElevenLabsApiConfiguration(), abortSignal, text, voiceId, modelId, voiceSettings, }) {
    return postJsonToApi({
        url: api.assembleUrl(`/text-to-speech/${voiceId}`),
        headers: api.headers,
        body: {
            text,
            model_id: modelId,
            voice_settings: voiceSettings != null
                ? {
                    stability: voiceSettings.stability,
                    similarity_boost: voiceSettings.similarityBoost,
                    style: voiceSettings.style,
                    use_speaker_boost: voiceSettings.useSpeakerBoost,
                }
                : undefined,
        },
        failedResponseHandler: failedElevenLabsCallResponseHandler,
        successfulResponseHandler: createAudioMpegResponseHandler(),
        abortSignal,
    });
}
