// modelfusion/model-function/generate-image/generateImage.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateImage = void 0;
const executeCall_js_1 = require("../executeCall.cjs");
const ImageGenerationPromise_js_1 = require("./ImageGenerationPromise.cjs");
/**
 * Generates a base64-encoded image using a prompt.
 * The prompt format depends on the model.
 * For example, OpenAI image models expect a string prompt,
 * and Stability AI models expect an array of text prompts with optional weights.
 *
 * @example
 * const image = await generateImage(
 *   new StabilityImageGenerationModel(...),
 *   [
 *     { text: "the wicked witch of the west" },
 *     { text: "style of early 19th century painting", weight: 0.5 },
 *   ]
 * );
 */
function generateImage(model, prompt, options) {
    return new ImageGenerationPromise_js_1.ImageGenerationPromise((0, executeCall_js_1.executeCall)({
        functionType: "image-generation",
        input: prompt,
        model,
        options,
        generateResponse: async (options) => {
            const result = await model.doGenerateImage(prompt, options);
            return {
                response: result.response,
                extractedValue: result.base64Image,
            };
        },
    }));
}
exports.generateImage = generateImage;
