// modelfusion/model-function/generate-image/ImageGenerationPromise.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageGenerationPromise = void 0;
class ImageGenerationPromise extends Promise {
    constructor(fullPromise) {
        super((resolve) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            resolve(null); // we override the resolve function
        });
        Object.defineProperty(this, "fullPromise", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: fullPromise
        });
        Object.defineProperty(this, "imageBase64Promise", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "imageBufferPromise", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.imageBase64Promise = fullPromise.then((result) => result.value);
        this.imageBufferPromise = this.imageBase64Promise.then((base64Text) => Buffer.from(base64Text, "base64"));
    }
    asFullResponse() {
        return this.fullPromise;
    }
    asBase64Text() {
        return this.imageBase64Promise;
    }
    asBuffer() {
        return this.imageBufferPromise;
    }
    then(onfulfilled, onrejected) {
        return this.imageBufferPromise.then(onfulfilled, onrejected);
    }
    catch(onrejected) {
        return this.imageBufferPromise.catch(onrejected);
    }
    finally(onfinally) {
        return this.imageBufferPromise.finally(onfinally);
    }
}
exports.ImageGenerationPromise = ImageGenerationPromise;
