// modelfusion/model-function/generate-image/PromptFormatImageGenerationModel.js
export class PromptFormatImageGenerationModel {
    constructor({ model, promptFormat, }) {
        Object.defineProperty(this, "model", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "promptFormat", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.model = model;
        this.promptFormat = promptFormat;
    }
    get modelInformation() {
        return this.model.modelInformation;
    }
    get settings() {
        return this.model.settings;
    }
    doGenerateImage(prompt, options) {
        const mappedPrompt = this.promptFormat.format(prompt);
        return this.model.doGenerateImage(mappedPrompt, options);
    }
    get settingsForEvent() {
        return this.model.settingsForEvent;
    }
    withPromptFormat(promptFormat) {
        return new PromptFormatImageGenerationModel({ model: this, promptFormat });
    }
    withSettings(additionalSettings) {
        return new PromptFormatImageGenerationModel({
            model: this.model.withSettings(additionalSettings),
            promptFormat: this.promptFormat,
        });
    }
}
