// modelfusion/model-provider/cohere/CohereTokenizer.js
import { z } from "zod";
import { callWithRetryAndThrottle } from "../../core/api/callWithRetryAndThrottle.js";
import { createJsonResponseHandler, postJsonToApi, } from "../../core/api/postToApi.js";
import { failedCohereCallResponseHandler } from "./CohereError.js";
import { CohereApiConfiguration } from "./CohereApiConfiguration.js";
/**
 * Tokenizer for the Cohere models. It uses the Co.Tokenize and Co.Detokenize APIs.
 *
 * @see https://docs.cohere.com/reference/tokenize
 * @see https://docs.cohere.com/reference/detokenize-1
 *
 * @example
 * const tokenizer = new CohereTokenizer({ model: "command-nightly" });
 *
 * const text = "At first, Nox didn't know what to do with the pup.";
 *
 * const tokenCount = await countTokens(tokenizer, text);
 * const tokens = await tokenizer.tokenize(text);
 * const tokensAndTokenTexts = await tokenizer.tokenizeWithTexts(text);
 * const reconstructedText = await tokenizer.detokenize(tokens);
 */
export class CohereTokenizer {
    constructor(settings) {
        Object.defineProperty(this, "settings", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.settings = settings;
    }
    async callTokenizeAPI(text, context) {
        return callWithRetryAndThrottle({
            retry: this.settings.api?.retry,
            throttle: this.settings.api?.throttle,
            call: async () => callCohereTokenizeAPI({
                abortSignal: context?.abortSignal,
                text,
                ...this.settings,
            }),
        });
    }
    async callDeTokenizeAPI(tokens, context) {
        return callWithRetryAndThrottle({
            retry: this.settings.api?.retry,
            throttle: this.settings.api?.throttle,
            call: async () => callCohereDetokenizeAPI({
                abortSignal: context?.abortSignal,
                tokens,
                ...this.settings,
            }),
        });
    }
    async tokenize(text) {
        return (await this.tokenizeWithTexts(text)).tokens;
    }
    async tokenizeWithTexts(text) {
        const response = await this.callTokenizeAPI(text);
        return {
            tokens: response.tokens,
            tokenTexts: response.token_strings,
        };
    }
    async detokenize(tokens) {
        const response = await this.callDeTokenizeAPI(tokens);
        return response.text;
    }
}
const cohereDetokenizationResponseSchema = z.object({
    text: z.string(),
    meta: z.object({
        api_version: z.object({
            version: z.string(),
        }),
    }),
});
async function callCohereDetokenizeAPI({ api = new CohereApiConfiguration(), abortSignal, model, tokens, }) {
    return postJsonToApi({
        url: api.assembleUrl(`/detokenize`),
        headers: api.headers,
        body: {
            model,
            tokens,
        },
        failedResponseHandler: failedCohereCallResponseHandler,
        successfulResponseHandler: createJsonResponseHandler(cohereDetokenizationResponseSchema),
        abortSignal,
    });
}
const cohereTokenizationResponseSchema = z.object({
    tokens: z.array(z.number()),
    token_strings: z.array(z.string()),
    meta: z.object({
        api_version: z.object({
            version: z.string(),
        }),
    }),
});
async function callCohereTokenizeAPI({ api = new CohereApiConfiguration(), abortSignal, model, text, }) {
    return postJsonToApi({
        url: api.assembleUrl(`/tokenize`),
        headers: api.headers,
        body: {
            model,
            text,
        },
        failedResponseHandler: failedCohereCallResponseHandler,
        successfulResponseHandler: createJsonResponseHandler(cohereTokenizationResponseSchema),
        abortSignal,
    });
}
