// modelfusion/model-provider/cohere/CohereTextEmbeddingModel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CohereTextEmbeddingModel = exports.COHERE_TEXT_EMBEDDING_MODELS = void 0;
const zod_1 = require("zod");
const callWithRetryAndThrottle_js_1 = require("../../core/api/callWithRetryAndThrottle.cjs");
const postToApi_js_1 = require("../../core/api/postToApi.cjs");
const AbstractModel_js_1 = require("../../model-function/AbstractModel.cjs");
const CohereApiConfiguration_js_1 = require("./CohereApiConfiguration.cjs");
const CohereError_js_1 = require("./CohereError.cjs");
const CohereTokenizer_js_1 = require("./CohereTokenizer.cjs");
exports.COHERE_TEXT_EMBEDDING_MODELS = {
    "embed-english-light-v2.0": {
        contextWindowSize: 4096,
        embeddingDimensions: 1024,
    },
    "embed-english-v2.0": {
        contextWindowSize: 4096,
        embeddingDimensions: 4096,
    },
    "embed-multilingual-v2.0": {
        contextWindowSize: 4096,
        embeddingDimensions: 768,
    },
};
/**
 * Create a text embedding model that calls the Cohere Co.Embed API.
 *
 * @see https://docs.cohere.com/reference/embed
 *
 * @example
 * const embeddings = await embedMany(
 *   new CohereTextEmbeddingModel({ model: "embed-english-light-v2.0" }),
 *   [
 *     "At first, Nox didn't know what to do with the pup.",
 *     "He keenly observed and absorbed everything around him, from the birds in the sky to the trees in the forest.",
 *   ]
 * );
 */
class CohereTextEmbeddingModel extends AbstractModel_js_1.AbstractModel {
    constructor(settings) {
        super({ settings });
        Object.defineProperty(this, "provider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "cohere"
        });
        Object.defineProperty(this, "maxValuesPerCall", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 96
        });
        Object.defineProperty(this, "embeddingDimensions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "contextWindowSize", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "tokenizer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.contextWindowSize =
            exports.COHERE_TEXT_EMBEDDING_MODELS[this.modelName].contextWindowSize;
        this.tokenizer = new CohereTokenizer_js_1.CohereTokenizer({
            api: this.settings.api,
            model: this.settings.model,
        });
        this.embeddingDimensions =
            exports.COHERE_TEXT_EMBEDDING_MODELS[this.modelName].embeddingDimensions;
    }
    get modelName() {
        return this.settings.model;
    }
    async tokenize(text) {
        return this.tokenizer.tokenize(text);
    }
    async tokenizeWithTexts(text) {
        return this.tokenizer.tokenizeWithTexts(text);
    }
    async detokenize(tokens) {
        return this.tokenizer.detokenize(tokens);
    }
    async callAPI(texts, options) {
        if (texts.length > this.maxValuesPerCall) {
            throw new Error(`The Cohere embedding API only supports ${this.maxValuesPerCall} texts per API call.`);
        }
        return (0, callWithRetryAndThrottle_js_1.callWithRetryAndThrottle)({
            retry: this.settings.api?.retry,
            throttle: this.settings.api?.throttle,
            call: async () => callCohereEmbeddingAPI({
                ...this.settings,
                abortSignal: options?.run?.abortSignal,
                texts,
            }),
        });
    }
    get settingsForEvent() {
        return {
            truncate: this.settings.truncate,
        };
    }
    async doEmbedValues(texts, options) {
        const response = await this.callAPI(texts, options);
        return {
            response,
            embeddings: response.embeddings,
        };
    }
    withSettings(additionalSettings) {
        return new CohereTextEmbeddingModel(Object.assign({}, this.settings, additionalSettings));
    }
}
exports.CohereTextEmbeddingModel = CohereTextEmbeddingModel;
const cohereTextEmbeddingResponseSchema = zod_1.z.object({
    id: zod_1.z.string(),
    texts: zod_1.z.array(zod_1.z.string()),
    embeddings: zod_1.z.array(zod_1.z.array(zod_1.z.number())),
    meta: zod_1.z.object({
        api_version: zod_1.z.object({
            version: zod_1.z.string(),
        }),
    }),
});
async function callCohereEmbeddingAPI({ api = new CohereApiConfiguration_js_1.CohereApiConfiguration(), abortSignal, model, texts, truncate, }) {
    return (0, postToApi_js_1.postJsonToApi)({
        url: api.assembleUrl(`/embed`),
        headers: api.headers,
        body: {
            model,
            texts,
            truncate,
        },
        failedResponseHandler: CohereError_js_1.failedCohereCallResponseHandler,
        successfulResponseHandler: (0, postToApi_js_1.createJsonResponseHandler)(cohereTextEmbeddingResponseSchema),
        abortSignal,
    });
}
