// modelfusion/model-provider/cohere/CohereError.js
import { z } from "zod";
import { ApiCallError } from "../../core/api/ApiCallError.js";
import { parseJsonWithZod } from "../../util/parseJSON.js";
export const cohereErrorDataSchema = z.object({
    message: z.string(),
});
export class CohereError extends ApiCallError {
    constructor({ data, statusCode, url, requestBodyValues, message = data.message, }) {
        super({ message, statusCode, requestBodyValues, url });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.data = data;
    }
}
export const failedCohereCallResponseHandler = async ({ response, url, requestBodyValues }) => {
    const responseBody = await response.text();
    // For some errors, the body of Cohere responses is empty:
    if (responseBody.trim() === "") {
        return new CohereError({
            url,
            requestBodyValues,
            statusCode: response.status,
            data: {
                message: response.statusText,
            },
        });
    }
    const parsedError = parseJsonWithZod(responseBody, cohereErrorDataSchema);
    return new CohereError({
        url,
        requestBodyValues,
        statusCode: response.status,
        data: parsedError,
    });
};
