// modelfusion/model-provider/openai/OpenAIImageGenerationModel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAIImageGenerationResponseFormat = exports.OpenAIImageGenerationModel = exports.calculateOpenAIImageGenerationCostInMillicents = void 0;
const zod_1 = require("zod");
const callWithRetryAndThrottle_js_1 = require("../../core/api/callWithRetryAndThrottle.cjs");
const postToApi_js_1 = require("../../core/api/postToApi.cjs");
const AbstractModel_js_1 = require("../../model-function/AbstractModel.cjs");
const PromptFormatImageGenerationModel_js_1 = require("../../model-function/generate-image/PromptFormatImageGenerationModel.cjs");
const OpenAIApiConfiguration_js_1 = require("./OpenAIApiConfiguration.cjs");
const OpenAIError_js_1 = require("./OpenAIError.cjs");
/**
 * @see https://openai.com/pricing
 */
const sizeToCostInMillicents = {
    "1024x1024": 2000,
    "512x512": 1800,
    "256x256": 1600,
};
const calculateOpenAIImageGenerationCostInMillicents = ({ settings, }) => (settings.n ?? 1) * sizeToCostInMillicents[settings.size ?? "1024x1024"];
exports.calculateOpenAIImageGenerationCostInMillicents = calculateOpenAIImageGenerationCostInMillicents;
/**
 * Create an image generation model that calls the OpenAI AI image creation API.
 *
 * @see https://platform.openai.com/docs/api-reference/images/create
 *
 * @example
 * const image = await generateImage(
 *   new OpenAIImageGenerationModel({ size: "512x512" }),
 *   "the wicked witch of the west in the style of early 19th century painting"
 * );
 */
class OpenAIImageGenerationModel extends AbstractModel_js_1.AbstractModel {
    constructor(settings) {
        super({ settings });
        Object.defineProperty(this, "provider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "openai"
        });
        Object.defineProperty(this, "modelName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: null
        });
    }
    async callAPI(prompt, options) {
        const run = options?.run;
        const responseFormat = options?.responseFormat;
        const callSettings = {
            ...this.settings,
            user: this.settings.isUserIdForwardingEnabled ? run?.userId : undefined,
            abortSignal: run?.abortSignal,
            responseFormat,
            prompt,
        };
        return (0, callWithRetryAndThrottle_js_1.callWithRetryAndThrottle)({
            retry: callSettings.api?.retry,
            throttle: callSettings.api?.throttle,
            call: async () => callOpenAIImageGenerationAPI(callSettings),
        });
    }
    get settingsForEvent() {
        const eventSettingProperties = [
            "n",
            "size",
        ];
        return Object.fromEntries(Object.entries(this.settings).filter(([key]) => eventSettingProperties.includes(key)));
    }
    async doGenerateImage(prompt, options) {
        const response = await this.callAPI(prompt, {
            responseFormat: exports.OpenAIImageGenerationResponseFormat.base64Json,
            ...options,
        });
        return {
            response,
            base64Image: response.data[0].b64_json,
        };
    }
    withPromptFormat(promptFormat) {
        return new PromptFormatImageGenerationModel_js_1.PromptFormatImageGenerationModel({
            model: this,
            promptFormat,
        });
    }
    withSettings(additionalSettings) {
        return new OpenAIImageGenerationModel(Object.assign({}, this.settings, additionalSettings));
    }
}
exports.OpenAIImageGenerationModel = OpenAIImageGenerationModel;
const openAIImageGenerationUrlSchema = zod_1.z.object({
    created: zod_1.z.number(),
    data: zod_1.z.array(zod_1.z.object({
        url: zod_1.z.string(),
    })),
});
const openAIImageGenerationBase64JsonSchema = zod_1.z.object({
    created: zod_1.z.number(),
    data: zod_1.z.array(zod_1.z.object({
        b64_json: zod_1.z.string(),
    })),
});
exports.OpenAIImageGenerationResponseFormat = {
    url: {
        type: "url",
        handler: (0, postToApi_js_1.createJsonResponseHandler)(openAIImageGenerationUrlSchema),
    },
    base64Json: {
        type: "b64_json",
        handler: (0, postToApi_js_1.createJsonResponseHandler)(openAIImageGenerationBase64JsonSchema),
    },
};
async function callOpenAIImageGenerationAPI({ api = new OpenAIApiConfiguration_js_1.OpenAIApiConfiguration(), abortSignal, prompt, n, size, responseFormat, user, }) {
    return (0, postToApi_js_1.postJsonToApi)({
        url: api.assembleUrl("/images/generations"),
        headers: api.headers,
        body: {
            prompt,
            n,
            size,
            response_format: responseFormat.type,
            user,
        },
        failedResponseHandler: OpenAIError_js_1.failedOpenAICallResponseHandler,
        successfulResponseHandler: responseFormat?.handler,
        abortSignal,
    });
}
