// modelfusion/model-function/generate-structure/generateStructure.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateStructure = void 0;
const executeCall_js_1 = require("../executeCall.cjs");
const ModelFunctionPromise_js_1 = require("../ModelFunctionPromise.cjs");
const StructureValidationError_js_1 = require("./StructureValidationError.cjs");
function generateStructure(model, structureDefinition, prompt, options) {
    // Note: PROMPT must not be a function.
    const expandedPrompt = typeof prompt === "function"
        ? prompt(structureDefinition)
        : prompt;
    return new ModelFunctionPromise_js_1.ModelFunctionPromise((0, executeCall_js_1.executeCall)({
        functionType: "structure-generation",
        input: expandedPrompt,
        model,
        options,
        generateResponse: async (options) => {
            const result = await model.doGenerateStructure(structureDefinition, expandedPrompt, options);
            const structure = result.value;
            const parseResult = structureDefinition.schema.validate(structure);
            if (!parseResult.success) {
                throw new StructureValidationError_js_1.StructureValidationError({
                    structureName: structureDefinition.name,
                    valueText: result.valueText,
                    value: structure,
                    cause: parseResult.error,
                });
            }
            const value = parseResult.data;
            return {
                response: result.response,
                extractedValue: value,
                usage: result.usage,
            };
        },
    }));
}
exports.generateStructure = generateStructure;
