// modelfusion/model-function/generate-structure/generateStructureOrText.js
import { executeCall } from "../executeCall.js";
import { ModelFunctionPromise } from "../ModelFunctionPromise.js";
import { NoSuchStructureError } from "./NoSuchStructureError.js";
import { StructureValidationError } from "./StructureValidationError.js";
export function generateStructureOrText(model, structureDefinitions, prompt, options) {
    // Note: PROMPT must not be a function.
    const expandedPrompt = typeof prompt === "function"
        ? prompt(structureDefinitions)
        : prompt;
    return new ModelFunctionPromise(executeCall({
        functionType: "structure-or-text-generation",
        input: expandedPrompt,
        model,
        options,
        generateResponse: async (options) => {
            const result = await model.doGenerateStructureOrText(structureDefinitions, expandedPrompt, options);
            const { structure, value, text } = result.structureAndText;
            // text generation:
            if (structure == null) {
                return {
                    response: result.response,
                    extractedValue: { structure, value, text },
                    usage: result.usage,
                };
            }
            const definition = structureDefinitions.find((d) => d.name === structure);
            if (definition == undefined) {
                throw new NoSuchStructureError(structure);
            }
            const parseResult = definition.schema.validate(value);
            if (!parseResult.success) {
                throw new StructureValidationError({
                    structureName: structure,
                    value,
                    valueText: result.structureAndText.valueText,
                    cause: parseResult.error,
                });
            }
            return {
                response: result.response,
                extractedValue: {
                    structure: structure,
                    value: parseResult.data,
                    text: text, // text is string | null, which is part of the response for schema values
                },
                usage: result.usage,
            };
        },
    }));
}
