// modelfusion/model-function/generate-structure/generateStructure.js
import { executeCall } from "../executeCall.js";
import { ModelFunctionPromise } from "../ModelFunctionPromise.js";
import { StructureValidationError } from "./StructureValidationError.js";
export function generateStructure(model, structureDefinition, prompt, options) {
    // Note: PROMPT must not be a function.
    const expandedPrompt = typeof prompt === "function"
        ? prompt(structureDefinition)
        : prompt;
    return new ModelFunctionPromise(executeCall({
        functionType: "structure-generation",
        input: expandedPrompt,
        model,
        options,
        generateResponse: async (options) => {
            const result = await model.doGenerateStructure(structureDefinition, expandedPrompt, options);
            const structure = result.value;
            const parseResult = structureDefinition.schema.validate(structure);
            if (!parseResult.success) {
                throw new StructureValidationError({
                    structureName: structureDefinition.name,
                    valueText: result.valueText,
                    value: structure,
                    cause: parseResult.error,
                });
            }
            const value = parseResult.data;
            return {
                response: result.response,
                extractedValue: value,
                usage: result.usage,
            };
        },
    }));
}
