// modelfusion/model-function/generate-structure/streamStructure.js
import { nanoid as createId } from "nanoid";
import { FunctionEventSource } from "../../core/FunctionEventSource.js";
import { getGlobalFunctionLogging } from "../../core/GlobalFunctionLogging.js";
import { getGlobalFunctionObservers } from "../../core/GlobalFunctionObservers.js";
import { AbortError } from "../../core/api/AbortError.js";
import { getFunctionCallLogger } from "../../core/getFunctionCallLogger.js";
import { getRun } from "../../core/getRun.js";
import { startDurationMeasurement } from "../../util/DurationMeasurement.js";
import { isDeepEqualData } from "../../util/isDeepEqualData.js";
import { runSafe } from "../../util/runSafe.js";
import { AsyncIterableResultPromise } from "../AsyncIterableResultPromise.js";
export function streamStructure(model, structureDefinition, prompt, options) {
    return new AsyncIterableResultPromise(doStreamStructure(model, structureDefinition, prompt, options));
}
async function doStreamStructure(model, structureDefinition, prompt, options) {
    const run = await getRun(options?.run);
    const settings = model.settings;
    const eventSource = new FunctionEventSource({
        observers: [
            ...getFunctionCallLogger(options?.logging ?? getGlobalFunctionLogging()),
            ...getGlobalFunctionObservers(),
            ...(settings.observers ?? []),
            ...(run?.functionObserver != null ? [run.functionObserver] : []),
            ...(options?.observers ?? []),
        ],
        errorHandler: run?.errorHandler,
    });
    const durationMeasurement = startDurationMeasurement();
    const startMetadata = {
        functionType: "structure-streaming",
        callId: `call-${createId()}`,
        runId: run?.runId,
        sessionId: run?.sessionId,
        userId: run?.userId,
        functionId: options?.functionId,
        model: model.modelInformation,
        settings: model.settingsForEvent,
        input: prompt,
        timestamp: durationMeasurement.startDate,
        startTimestamp: durationMeasurement.startDate,
    };
    eventSource.notify({
        eventType: "started",
        ...startMetadata,
    });
    const result = await runSafe(async () => {
        const deltaIterable = await model.doStreamStructure(structureDefinition, prompt, {
            functionId: options?.functionId,
            logging: options?.logging,
            observers: options?.observers,
            run,
        });
        return (async function* () {
            function reportError(error) {
                const finishMetadata = {
                    eventType: "finished",
                    ...startMetadata,
                    finishTimestamp: new Date(),
                    durationInMs: durationMeasurement.durationInMs,
                };
                eventSource.notify(error instanceof AbortError
                    ? {
                        ...finishMetadata,
                        result: {
                            status: "abort",
                        },
                    }
                    : {
                        ...finishMetadata,
                        result: {
                            status: "error",
                            error,
                        },
                    });
            }
            let lastStructure;
            let lastFullDelta;
            for await (const event of deltaIterable) {
                if (event?.type === "error") {
                    reportError(event.error);
                    throw event.error;
                }
                if (event?.type === "delta") {
                    const latestFullDelta = event.fullDelta;
                    const latestStructure = event.valueDelta;
                    // only send a new part into the stream when the partial structure has changed:
                    if (!isDeepEqualData(lastStructure, latestStructure)) {
                        lastFullDelta = latestFullDelta;
                        lastStructure = latestStructure;
                        yield {
                            isComplete: false,
                            value: lastStructure,
                        };
                    }
                }
            }
            // process the final result (full type validation):
            const parseResult = structureDefinition.schema.validate(lastStructure);
            if (!parseResult.success) {
                reportError(parseResult.error);
                throw parseResult.error;
            }
            yield {
                isComplete: true,
                value: parseResult.data,
            };
            const finishMetadata = {
                eventType: "finished",
                ...startMetadata,
                finishTimestamp: new Date(),
                durationInMs: durationMeasurement.durationInMs,
            };
            eventSource.notify({
                ...finishMetadata,
                result: {
                    status: "success",
                    response: lastFullDelta,
                    value: lastStructure,
                },
            });
        })();
    });
    if (!result.ok) {
        const finishMetadata = {
            eventType: "finished",
            ...startMetadata,
            finishTimestamp: new Date(),
            durationInMs: durationMeasurement.durationInMs,
        };
        if (result.isAborted) {
            eventSource.notify({
                ...finishMetadata,
                result: {
                    status: "abort",
                },
            });
            throw new AbortError();
        }
        eventSource.notify({
            ...finishMetadata,
            result: {
                status: "error",
                error: result.error,
            },
        });
        throw result.error;
    }
    return {
        output: result.value,
        metadata: startMetadata,
    };
}
