// modelfusion/model-function/generate-structure/StructureFromTextGenerationModel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StructureFromTextGenerationModel = void 0;
const generateText_js_1 = require("../generate-text/generateText.cjs");
const StructureParseError_js_1 = require("./StructureParseError.cjs");
class StructureFromTextGenerationModel {
    constructor({ model, format, }) {
        Object.defineProperty(this, "model", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "format", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.model = model;
        this.format = format;
    }
    get modelInformation() {
        return this.model.modelInformation;
    }
    get settings() {
        return this.model.settings;
    }
    get settingsForEvent() {
        return this.model.settingsForEvent;
    }
    async doGenerateStructure(structure, prompt, options) {
        const { response, value } = await (0, generateText_js_1.generateText)(this.model, this.format.createPrompt(prompt, structure), options).asFullResponse();
        try {
            return {
                response,
                value: this.format.extractStructure(value),
                valueText: value,
            };
        }
        catch (error) {
            throw new StructureParseError_js_1.StructureParseError({
                structureName: structure.name,
                valueText: value,
                cause: error,
            });
        }
    }
    withSettings(additionalSettings) {
        return new StructureFromTextGenerationModel({
            model: this.model.withSettings(additionalSettings),
            format: this.format,
        });
    }
}
exports.StructureFromTextGenerationModel = StructureFromTextGenerationModel;
