// modelfusion/model-provider/huggingface/HuggingFaceImageDescriptionModel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HuggingFaceImageDescriptionModel = void 0;
const zod_1 = require("zod");
const callWithRetryAndThrottle_js_1 = require("../../core/api/callWithRetryAndThrottle.cjs");
const postToApi_js_1 = require("../../core/api/postToApi.cjs");
const AbstractModel_js_1 = require("../../model-function/AbstractModel.cjs");
const HuggingFaceApiConfiguration_js_1 = require("./HuggingFaceApiConfiguration.cjs");
const HuggingFaceError_js_1 = require("./HuggingFaceError.cjs");
/**
 * Create an image to text model that calls a Hugging Face Image-to-Text Inference API.
 *
 * @see https://huggingface.co/tasks/image-to-text
 */
class HuggingFaceImageDescriptionModel extends AbstractModel_js_1.AbstractModel {
    constructor(settings) {
        super({ settings });
        Object.defineProperty(this, "provider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "huggingface"
        });
        Object.defineProperty(this, "countPromptTokens", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: undefined
        });
    }
    get modelName() {
        return this.settings.model;
    }
    async callAPI(data, options) {
        return (0, callWithRetryAndThrottle_js_1.callWithRetryAndThrottle)({
            retry: this.settings.api?.retry,
            throttle: this.settings.api?.throttle,
            call: async () => callHuggingFaceImageDescriptionAPI({
                ...this.settings,
                abortSignal: options?.run?.abortSignal,
                data,
            }),
        });
    }
    get settingsForEvent() {
        return {};
    }
    async doDescribeImage(data, options) {
        const response = await this.callAPI(data, options);
        return {
            response,
            description: response[0].generated_text,
        };
    }
    withSettings(additionalSettings) {
        return new HuggingFaceImageDescriptionModel(Object.assign({}, this.settings, additionalSettings));
    }
}
exports.HuggingFaceImageDescriptionModel = HuggingFaceImageDescriptionModel;
const huggingFaceImageDescriptionResponseSchema = zod_1.z.array(zod_1.z.object({
    generated_text: zod_1.z.string(),
}));
async function callHuggingFaceImageDescriptionAPI({ api = new HuggingFaceApiConfiguration_js_1.HuggingFaceApiConfiguration(), abortSignal, model, data, }) {
    return (0, postToApi_js_1.postToApi)({
        url: api.assembleUrl(`/${model}`),
        headers: api.headers,
        body: {
            content: data,
            values: {},
        },
        failedResponseHandler: HuggingFaceError_js_1.failedHuggingFaceCallResponseHandler,
        successfulResponseHandler: (0, postToApi_js_1.createJsonResponseHandler)(huggingFaceImageDescriptionResponseSchema),
        abortSignal,
    });
}
