// modelfusion/model-provider/huggingface/HuggingFaceError.js
import { z } from "zod";
import { ApiCallError } from "../../core/api/ApiCallError.js";
import { parseJsonWithZod } from "../../util/parseJSON.js";
export const huggingFaceErrorDataSchema = z.object({
    error: z.array(z.string()).or(z.string()),
});
export class HuggingFaceError extends ApiCallError {
    constructor({ data, statusCode, url, requestBodyValues, message = typeof data.error === "string"
        ? data.error
        : data.error.join("\n\n"), }) {
        super({ message, statusCode, requestBodyValues, url });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.data = data;
    }
}
export const failedHuggingFaceCallResponseHandler = async ({ response, url, requestBodyValues }) => new HuggingFaceError({
    url,
    requestBodyValues,
    statusCode: response.status,
    data: parseJsonWithZod(await response.text(), huggingFaceErrorDataSchema),
});
