// modelfusion/model-provider/huggingface/HuggingFaceError.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.failedHuggingFaceCallResponseHandler = exports.HuggingFaceError = exports.huggingFaceErrorDataSchema = void 0;
const zod_1 = require("zod");
const ApiCallError_js_1 = require("../../core/api/ApiCallError.cjs");
const parseJSON_js_1 = require("../../util/parseJSON.cjs");
exports.huggingFaceErrorDataSchema = zod_1.z.object({
    error: zod_1.z.array(zod_1.z.string()).or(zod_1.z.string()),
});
class HuggingFaceError extends ApiCallError_js_1.ApiCallError {
    constructor({ data, statusCode, url, requestBodyValues, message = typeof data.error === "string"
        ? data.error
        : data.error.join("\n\n"), }) {
        super({ message, statusCode, requestBodyValues, url });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.data = data;
    }
}
exports.HuggingFaceError = HuggingFaceError;
const failedHuggingFaceCallResponseHandler = async ({ response, url, requestBodyValues }) => new HuggingFaceError({
    url,
    requestBodyValues,
    statusCode: response.status,
    data: (0, parseJSON_js_1.parseJsonWithZod)(await response.text(), exports.huggingFaceErrorDataSchema),
});
exports.failedHuggingFaceCallResponseHandler = failedHuggingFaceCallResponseHandler;
