// modelfusion/model-provider/lmnt/LmntSpeechSynthesisModel.js
import { AbstractModel } from "../../model-function/AbstractModel.js";
import { callWithRetryAndThrottle } from "../../core/api/callWithRetryAndThrottle.js";
import { createAudioMpegResponseHandler, postToApi, } from "../../core/api/postToApi.js";
import { LmntApiConfiguration } from "./LmntApiConfiguration.js";
import { failedLmntCallResponseHandler } from "./LmntError.js";
/**
 * Synthesize speech using the LMNT API.
 *
 * @see https://www.lmnt.com/docs/rest/#synthesize-speech
 */
export class LmntSpeechSynthesisModel extends AbstractModel {
    constructor(settings) {
        super({ settings });
        Object.defineProperty(this, "provider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "lmnt"
        });
    }
    get modelName() {
        return this.settings.voice;
    }
    async callAPI(text, options) {
        return callWithRetryAndThrottle({
            retry: this.settings.api?.retry,
            throttle: this.settings.api?.throttle,
            call: async () => callLmntTextToSpeechAPI({
                ...this.settings,
                abortSignal: options?.run?.abortSignal,
                text,
            }),
        });
    }
    get settingsForEvent() {
        return {
            voice: this.settings.voice,
            speed: this.settings.speed,
            seed: this.settings.seed,
            length: this.settings.length,
        };
    }
    generateSpeechResponse(text, options) {
        return this.callAPI(text, options);
    }
    withSettings(additionalSettings) {
        return new LmntSpeechSynthesisModel({
            ...this.settings,
            ...additionalSettings,
        });
    }
}
async function callLmntTextToSpeechAPI({ api = new LmntApiConfiguration(), abortSignal, text, voice, speed, seed, length, }) {
    const formData = new FormData();
    formData.append("text", text);
    formData.append("voice", voice);
    formData.append("format", "mp3");
    if (speed != null)
        formData.append("speed", speed.toString());
    if (seed != null)
        formData.append("seed", seed.toString());
    if (length != null)
        formData.append("length", length.toString());
    return postToApi({
        url: api.assembleUrl(`/synthesize`),
        headers: api.headers,
        body: {
            content: formData,
            values: {
                text,
                voice,
                speed,
                seed,
                length,
            },
        },
        failedResponseHandler: failedLmntCallResponseHandler,
        successfulResponseHandler: createAudioMpegResponseHandler(),
        abortSignal,
    });
}
