// modelfusion/tool/executeTool.js
import { nanoid as createId } from "nanoid";
import { FunctionEventSource } from "../core/FunctionEventSource.js";
import { getGlobalFunctionLogging } from "../core/GlobalFunctionLogging.js";
import { getGlobalFunctionObservers } from "../core/GlobalFunctionObservers.js";
import { AbortError } from "../core/api/AbortError.js";
import { getFunctionCallLogger } from "../core/getFunctionCallLogger.js";
import { getRun } from "../core/getRun.js";
import { startDurationMeasurement } from "../util/DurationMeasurement.js";
import { runSafe } from "../util/runSafe.js";
import { ToolExecutionError } from "./ToolExecutionError.js";
export class ExecuteToolPromise extends Promise {
    constructor(fullPromise) {
        super((resolve) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            resolve(null); // we override the resolve function
        });
        Object.defineProperty(this, "fullPromise", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: fullPromise
        });
        Object.defineProperty(this, "outputPromise", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.outputPromise = fullPromise.then((result) => result.output);
    }
    asFullResponse() {
        return this.fullPromise;
    }
    then(onfulfilled, onrejected) {
        return this.outputPromise.then(onfulfilled, onrejected);
    }
    catch(onrejected) {
        return this.outputPromise.catch(onrejected);
    }
    finally(onfinally) {
        return this.outputPromise.finally(onfinally);
    }
}
/**
 * `executeTool` directly executes a tool with the given parameters.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function executeTool(tool, input, options) {
    return new ExecuteToolPromise(doExecuteTool(tool, input, options));
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function doExecuteTool(tool, input, options) {
    const run = await getRun(options?.run);
    const eventSource = new FunctionEventSource({
        observers: [
            ...getFunctionCallLogger(options?.logging ?? getGlobalFunctionLogging()),
            ...getGlobalFunctionObservers(),
            ...(run?.functionObserver != null ? [run.functionObserver] : []),
            ...(options?.observers ?? []),
        ],
        errorHandler: run?.errorHandler,
    });
    const durationMeasurement = startDurationMeasurement();
    const metadata = {
        functionType: "execute-tool",
        callId: `call-${createId()}`,
        runId: run?.runId,
        sessionId: run?.sessionId,
        userId: run?.userId,
        functionId: options?.functionId,
        toolName: tool.name,
        input,
    };
    eventSource.notify({
        ...metadata,
        eventType: "started",
        timestamp: durationMeasurement.startDate,
        startTimestamp: durationMeasurement.startDate,
    });
    const result = await runSafe(() => tool.execute(input, {
        functionId: options?.functionId,
        logging: options?.logging,
        observers: options?.observers,
        run,
    }));
    const finishMetadata = {
        ...metadata,
        eventType: "finished",
        timestamp: new Date(),
        startTimestamp: durationMeasurement.startDate,
        finishTimestamp: new Date(),
        durationInMs: durationMeasurement.durationInMs,
    };
    if (!result.ok) {
        if (result.isAborted) {
            eventSource.notify({
                ...finishMetadata,
                result: {
                    status: "abort",
                },
            });
            throw new AbortError();
        }
        eventSource.notify({
            ...finishMetadata,
            result: {
                status: "error",
                error: result.error,
            },
        });
        throw new ToolExecutionError({
            toolName: tool.name,
            input,
            cause: result.error,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            message: result.error?.message,
        });
    }
    const output = result.value;
    eventSource.notify({
        ...finishMetadata,
        result: {
            status: "success",
            value: output,
        },
    });
    return {
        output,
        metadata: finishMetadata,
    };
}
