// modelfusion/tool/useToolOrGenerateText.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useToolOrGenerateText = void 0;
const generateStructureOrText_js_1 = require("../model-function/generate-structure/generateStructureOrText.cjs");
const NoSuchToolError_js_1 = require("./NoSuchToolError.cjs");
const executeTool_js_1 = require("./executeTool.cjs");
async function useToolOrGenerateText(model, tools, prompt, options) {
    // Note: PROMPT must not be a function.
    const expandedPrompt = typeof prompt === "function"
        ? prompt(tools)
        : prompt;
    const modelResponse = await (0, generateStructureOrText_js_1.generateStructureOrText)(model, tools.map((tool) => ({
        name: tool.name,
        description: tool.description,
        schema: tool.inputSchema,
    })), expandedPrompt, options);
    const { structure, text } = modelResponse;
    if (structure == null) {
        return {
            tool: null,
            parameters: null,
            result: null,
            text,
        };
    }
    const tool = tools.find((tool) => tool.name === structure);
    if (tool == null) {
        throw new NoSuchToolError_js_1.NoSuchToolError(structure.toString());
    }
    const toolParameters = modelResponse.value;
    const result = await (0, executeTool_js_1.executeTool)(tool, toolParameters, options);
    return {
        tool: structure,
        result,
        parameters: toolParameters,
        text: text, // string | null is the expected value here
    };
}
exports.useToolOrGenerateText = useToolOrGenerateText;
