// modelfusion/tool/useTool.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTool = void 0;
const generateStructure_js_1 = require("../model-function/generate-structure/generateStructure.cjs");
const executeTool_js_1 = require("./executeTool.cjs");
/**
 * `useTool` uses `generateStructure` to generate parameters for a tool and then executes the tool with the parameters.
 *
 * @returns The result contains the name of the tool (`tool` property),
 * the parameters (`parameters` property, typed),
 * and the result of the tool execution (`result` property, typed).
 */
async function useTool(model, tool, prompt, options) {
    // Note: PROMPT must not be a function.
    const expandedPrompt = typeof prompt === "function"
        ? prompt(tool)
        : prompt;
    const { value } = await (0, generateStructure_js_1.generateStructure)(model, {
        name: tool.name,
        description: tool.description,
        schema: tool.inputSchema,
    }, expandedPrompt, options).asFullResponse();
    return {
        tool: tool.name,
        parameters: value,
        result: await (0, executeTool_js_1.executeTool)(tool, value, options),
    };
}
exports.useTool = useTool;
