// modelfusion/model-provider/automatic1111/Automatic1111Error.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.failedAutomatic1111CallResponseHandler = exports.Automatic1111Error = exports.automatic1111ErrorDataSchema = void 0;
const zod_1 = require("zod");
const ApiCallError_js_1 = require("../../core/api/ApiCallError.cjs");
const parseJSON_js_1 = require("../../util/parseJSON.cjs");
exports.automatic1111ErrorDataSchema = zod_1.z.object({
    error: zod_1.z.string(),
    detail: zod_1.z.string(),
    body: zod_1.z.string(),
    errors: zod_1.z.string(),
});
class Automatic1111Error extends ApiCallError_js_1.ApiCallError {
    constructor({ data, statusCode, url, requestBodyValues, message = data.detail, }) {
        super({ message, statusCode, requestBodyValues, url });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.data = data;
    }
}
exports.Automatic1111Error = Automatic1111Error;
const failedAutomatic1111CallResponseHandler = async ({ response, url, requestBodyValues }) => {
    const responseBody = await response.text();
    const parsedError = (0, parseJSON_js_1.parseJsonWithZod)(responseBody, exports.automatic1111ErrorDataSchema);
    return new Automatic1111Error({
        url,
        requestBodyValues,
        statusCode: response.status,
        data: parsedError,
    });
};
exports.failedAutomatic1111CallResponseHandler = failedAutomatic1111CallResponseHandler;
