// modelfusion/model-provider/automatic1111/Automatic1111ImageGenerationModel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Automatic1111ImageGenerationModel = void 0;
const zod_1 = require("zod");
const callWithRetryAndThrottle_js_1 = require("../../core/api/callWithRetryAndThrottle.cjs");
const postToApi_js_1 = require("../../core/api/postToApi.cjs");
const AbstractModel_js_1 = require("../../model-function/AbstractModel.cjs");
const PromptFormatImageGenerationModel_js_1 = require("../../model-function/generate-image/PromptFormatImageGenerationModel.cjs");
const Automatic1111ApiConfiguration_js_1 = require("./Automatic1111ApiConfiguration.cjs");
const Automatic1111Error_js_1 = require("./Automatic1111Error.cjs");
const Automatic1111ImageGenerationPrompt_js_1 = require("./Automatic1111ImageGenerationPrompt.cjs");
/**
 * Create an image generation model that calls the AUTOMATIC1111 Stable Diffusion Web UI API.
 *
 * @see https://github.com/AUTOMATIC1111/stable-diffusion-webui
 */
class Automatic1111ImageGenerationModel extends AbstractModel_js_1.AbstractModel {
    constructor(settings) {
        super({ settings });
        Object.defineProperty(this, "provider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "Automatic1111"
        });
    }
    get modelName() {
        return this.settings.model;
    }
    async callAPI(input, options) {
        return (0, callWithRetryAndThrottle_js_1.callWithRetryAndThrottle)({
            retry: this.settings.api?.retry,
            throttle: this.settings.api?.throttle,
            call: async () => callAutomatic1111ImageGenerationAPI({
                ...this.settings,
                abortSignal: options?.run?.abortSignal,
                prompt: input.prompt,
            }),
        });
    }
    get settingsForEvent() {
        return {
            height: this.settings.height,
            width: this.settings.width,
            sampler: this.settings.sampler,
            steps: this.settings.steps,
        };
    }
    async doGenerateImage(prompt, options) {
        const response = await this.callAPI(prompt, options);
        return {
            response,
            base64Image: response.images[0],
        };
    }
    withBasicPrompt() {
        return this.withPromptFormat((0, Automatic1111ImageGenerationPrompt_js_1.mapBasicPromptToAutomatic1111Format)());
    }
    withPromptFormat(promptFormat) {
        return new PromptFormatImageGenerationModel_js_1.PromptFormatImageGenerationModel({
            model: this,
            promptFormat,
        });
    }
    withSettings(additionalSettings) {
        return new Automatic1111ImageGenerationModel(Object.assign({}, this.settings, additionalSettings));
    }
}
exports.Automatic1111ImageGenerationModel = Automatic1111ImageGenerationModel;
const Automatic1111ImageGenerationResponseSchema = zod_1.z.object({
    images: zod_1.z.array(zod_1.z.string()),
    parameters: zod_1.z.object({}),
    info: zod_1.z.string(),
});
async function callAutomatic1111ImageGenerationAPI({ api = new Automatic1111ApiConfiguration_js_1.Automatic1111ApiConfiguration(), abortSignal, height, width, prompt, negativePrompt, sampler, steps, seed, model, }) {
    return (0, postToApi_js_1.postJsonToApi)({
        url: api.assembleUrl(`/txt2img`),
        headers: api.headers,
        body: {
            height,
            width,
            prompt,
            negative_prompt: negativePrompt,
            sampler_index: sampler,
            steps,
            seed,
            override_settings: {
                sd_model_checkpoint: model,
            },
        },
        failedResponseHandler: Automatic1111Error_js_1.failedAutomatic1111CallResponseHandler,
        successfulResponseHandler: (0, postToApi_js_1.createJsonResponseHandler)(Automatic1111ImageGenerationResponseSchema),
        abortSignal,
    });
}
