// modelfusion/model-function/generate-text/PromptFormatTextStreamingModel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptFormatTextStreamingModel = void 0;
const PromptFormatTextGenerationModel_js_1 = require("./PromptFormatTextGenerationModel.cjs");
class PromptFormatTextStreamingModel extends PromptFormatTextGenerationModel_js_1.PromptFormatTextGenerationModel {
    constructor(options) {
        super(options);
    }
    doStreamText(prompt, options) {
        const mappedPrompt = this.promptFormat.format(prompt);
        return this.model.doStreamText(mappedPrompt, options);
    }
    withPromptFormat(promptFormat) {
        return new PromptFormatTextStreamingModel({
            model: this.withSettings({
                stopSequences: [
                    ...(this.settings.stopSequences ?? []),
                    ...promptFormat.stopSequences,
                ],
            }),
            promptFormat,
        });
    }
    withSettings(additionalSettings) {
        return new PromptFormatTextStreamingModel({
            model: this.model.withSettings(additionalSettings),
            promptFormat: this.promptFormat,
        });
    }
}
exports.PromptFormatTextStreamingModel = PromptFormatTextStreamingModel;
