// modelfusion/model-function/generate-text/PromptFormatTextGenerationModel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptFormatTextGenerationModel = void 0;
class PromptFormatTextGenerationModel {
    constructor({ model, promptFormat, }) {
        Object.defineProperty(this, "model", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "promptFormat", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.model = model;
        this.promptFormat = promptFormat;
    }
    get modelInformation() {
        return this.model.modelInformation;
    }
    get settings() {
        return this.model.settings;
    }
    get tokenizer() {
        return this.model.tokenizer;
    }
    get contextWindowSize() {
        return this.model.contextWindowSize;
    }
    get countPromptTokens() {
        const originalCountPromptTokens = this.model.countPromptTokens?.bind(this.model);
        if (originalCountPromptTokens === undefined) {
            return undefined;
        }
        return ((prompt) => originalCountPromptTokens(this.promptFormat.format(prompt)));
    }
    doGenerateText(prompt, options) {
        const mappedPrompt = this.promptFormat.format(prompt);
        return this.model.doGenerateText(mappedPrompt, options);
    }
    get settingsForEvent() {
        return this.model.settingsForEvent;
    }
    withPromptFormat(promptFormat) {
        return new PromptFormatTextGenerationModel({
            model: this.withSettings({
                stopSequences: [
                    ...(this.settings.stopSequences ?? []),
                    ...promptFormat.stopSequences,
                ],
            }),
            promptFormat,
        });
    }
    withSettings(additionalSettings) {
        return new PromptFormatTextGenerationModel({
            model: this.model.withSettings(additionalSettings),
            promptFormat: this.promptFormat,
        });
    }
}
exports.PromptFormatTextGenerationModel = PromptFormatTextGenerationModel;
