// modelfusion/model-provider/llamacpp/LlamaCppError.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.failedLlamaCppCallResponseHandler = exports.LlamaCppError = exports.llamaCppErrorDataSchema = void 0;
const zod_1 = require("zod");
const ApiCallError_js_1 = require("../../core/api/ApiCallError.cjs");
const parseJSON_js_1 = require("../../util/parseJSON.cjs");
exports.llamaCppErrorDataSchema = zod_1.z.object({
    error: zod_1.z.string(),
});
class LlamaCppError extends ApiCallError_js_1.ApiCallError {
    constructor({ data, statusCode, url, requestBodyValues, message = data.error, }) {
        super({ message, statusCode, requestBodyValues, url });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.data = data;
    }
}
exports.LlamaCppError = LlamaCppError;
const failedLlamaCppCallResponseHandler = async ({ response, url, requestBodyValues }) => new LlamaCppError({
    url,
    requestBodyValues,
    statusCode: response.status,
    data: (0, parseJSON_js_1.parseJsonWithZod)(await response.text(), exports.llamaCppErrorDataSchema),
});
exports.failedLlamaCppCallResponseHandler = failedLlamaCppCallResponseHandler;
