// modelfusion/model-provider/llamacpp/LlamaCppTokenizer.js
import { z } from "zod";
import { callWithRetryAndThrottle } from "../../core/api/callWithRetryAndThrottle.js";
import { createJsonResponseHandler, postJsonToApi, } from "../../core/api/postToApi.js";
import { LlamaCppApiConfiguration } from "./LlamaCppApiConfiguration.js";
import { failedLlamaCppCallResponseHandler } from "./LlamaCppError.js";
/**
 * Tokenizer for LlamaCpp.

 * @example
 * const tokenizer = new LlamaCppTokenizer();
 *
 * const text = "At first, Nox didn't know what to do with the pup.";
 *
 * const tokenCount = await countTokens(tokenizer, text);
 * const tokens = await tokenizer.tokenize(text);
 * const tokensAndTokenTexts = await tokenizer.tokenizeWithTexts(text);
 * const reconstructedText = await tokenizer.detokenize(tokens);
 */
export class LlamaCppTokenizer {
    constructor(api = new LlamaCppApiConfiguration()) {
        Object.defineProperty(this, "api", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.api = api;
    }
    async callTokenizeAPI(text, context) {
        return callWithRetryAndThrottle({
            retry: this.api.retry,
            throttle: this.api.throttle,
            call: async () => callLlamaCppTokenizeAPI({
                api: this.api,
                abortSignal: context?.abortSignal,
                text,
            }),
        });
    }
    async tokenize(text) {
        const response = await this.callTokenizeAPI(text);
        return response.tokens;
    }
}
const llamaCppTokenizationResponseSchema = z.object({
    tokens: z.array(z.number()),
});
async function callLlamaCppTokenizeAPI({ api, abortSignal, text, }) {
    return postJsonToApi({
        url: api.assembleUrl(`/tokenize`),
        headers: api.headers,
        body: {
            content: text,
        },
        failedResponseHandler: failedLlamaCppCallResponseHandler,
        successfulResponseHandler: createJsonResponseHandler(llamaCppTokenizationResponseSchema),
        abortSignal,
    });
}
